### ISSUE TITLE
Changes on `mso_schema_site_anp_epg_domain` are not reflected on the NDO/APIC controllers.

### ISSUE THREAD

**Issue by edudppaz** [CiscoDevNet/terraform-provider-mso#197]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

* Terraform v1.4.5

### MSO version

* Version 2.3(1c)

### APIC version and APIC Platform for Site Level Resources

* V 5.2(7f) and on-prem

### Affected Resource(s)

* mso_schema_site_anp_epg_domain

### Terraform Configuration Files

```hcl
resource "mso_schema_site_anp_epg_domain" "epg_domain_l2_sf" {
  for_each    = { for bd in  local.tn_bd_flatten : bd.bd_name => bd } # This is a map that contains the tenant/bd/epg/domain info
  schema_id                = mso_schema.schemas[each.value.tn_name].id # Two schemas
  template_name            = "L2-STRETCHED" # This is a template associated with two sites
  site_id                  = data.mso_site.ndo_sites["San Francisco"].id # Only one site for this test. However, it is applied to both sites
  anp_name                 = "L2-LEGACY" # An application profile associated with the EPGs
  epg_name                 = each.value.epg_name
  domain_type              = "physicalDomain"
  domain_name              = "phy-dom"
  deploy_immediacy         = "immediate"
  resolution_immediacy     = "pre-provision"
}
```

### Expected Behavior

- The resource should push the domain association  into the EPG configuration on the selected site. 
- Re-running terraform should not do any changes
- Destroying the resource should remove the domain association

### Actual Behavior

- Changing the domain or pushing a new domain, does not remove the previous associated domain
- Destroying the resource does not remove the associated domain
- If the domain is changed, it associates the new domain in addition to the previous one
- Trying to push back the old domain, it gives an error that the domain is duplicated.

Seems like idempotency is missing from this resource.

### Steps to Reproduce

1. `terraform apply`
2. Try terraform destroy and/or:
3. Change the domain_name and verify on the APIC GUI that both domains are associated, the previous one is not removed.

### References

* #0000

**Comment by edudppaz** [comment]
As an extra comment, this is a problem as im trying to use the `mso_schema_template_anp_epg` resource to create the EPGs, and the `aci_physical_domain` to associate a domain with the EPG.

However, it seems like `mso_schema_template_anp_epg` does a change on the EPG that is in conflict with the domain association, as each time i want to run my terraform modules after deploying the templates, the `aci_physical_domain` module needs to re-create the domain association, creating downtime on the fabric.

If i run the `aci_physical_domain` multiple times without re-deploying the template, there is no change on the infrastructure.

**Comment by edudppaz** [comment]
Some more information.

How to reproduce:
- physical domain association is pushed into EPGs on two sites via NDO with the `mso_schema_site_anp_epg_domain` resource
- after the deployment, the physical domain association was destroyed in terraform by removing the `mso_schema_site_anp_epg_domain` resource
- NDO and subsequently, the APIC still shows the physical domain attached

Even after destroying the resource, NDO stores  the physical domain association on the template, re-deploying the template, effectively re-deploys the physical domain association every time, while terraform shows that "no changes are being made".
The terraform states shows no reference to the physical domain being changed by the resource, therefore there is a missmatch  between the real state on NDO and the terraform state.

Terraform output before applying (only deploying), shows no changes to the domain association:

```hcl
Terraform will perform the following actions:

  # module.tf-aci-ndo-deploy.mso_schema_template_deploy_ndo.template_deployer_l2["6448df03bd877951fbeefadf"] will be updated in-place
  ~ resource "mso_schema_template_deploy_ndo" "template_deployer_l2" {
      + force_apply   = "always-deploy"
        id            = "6448df03bd877951fbeefadf"
        # (3 unchanged attributes hidden)
    }

  # module.tf-aci-ndo-deploy.mso_schema_template_deploy_ndo.template_deployer_l2["6448df03bd877951fbeefae3"] will be updated in-place
  ~ resource "mso_schema_template_deploy_ndo" "template_deployer_l2" {
      + force_apply   = "always-deploy"
        id            = "6448df03bd877951fbeefae3"
        # (3 unchanged attributes hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy.
```

ACI EPG configuration shows no domain association before deploying.

After the deploy, the ACI EPG configuration shows:

```json
{
    "totalCount": "1",
    "imdata": [
        {
            "fvAEPg": {
                "children": [
{
                        "fvRsDomAtt": {
                            "attributes": {
                                "rn": "rsdomAtt-[uni/phys-phy-dom]",
...
```

This association comes from the NDO deployment even after the configuration was destroyed by terraform and is no longer showing as a change in terraform.
NDO GUI shows the phy-dom associated to the site. This is not reflected in the terraform state or resource.

NDO diff:

```diff

  "anps": [
    {
      "anpRef": "/schemas/6448df03bd877951fbeefae3/templates/L2-STRETCHED/anps/L2-LEGACY",
      "epgs": [
        {
-          "domainAssociations": [],
+          "domainAssociations": [
+            {
+              "allowMicroSegmentation": false,
+              "deployImmediacy": "lazy",
+              "dn": "uni/phys-phy-dom",
+              "domainType": "physicalDomain",
+              "resolutionImmediacy": "pre-provision"
+            }
+          ],
```

**Comment by edudppaz** [comment]
Could this be a commit that is still not in the code? I just started debugging terraform and i can see that the resource just exits without any error when trying to destroy.

However, i just compiled the provider directly from source (with some extra print statements for debugging) and it works as expected. So it seems to be something different between the version on the terraform registry (v0.9.0) and the main branch on source

**Comment by edudppaz** [comment]
Found it. I couldnt find any bug on the code, but it seems like there was an fix on commit 93142ce  by @akinross  (already merged into main on PR #180 ) which fixed an error on the indexCount using a wrong variable. It seems like this not only impacted #178 , but also the general behaviour of the resource when destroying or updating.

```diff
-	index, err := indexCount(cont, siteId, anpName, epgName, id)
+	index, err := indexCount(cont, siteId, anpName, epgName, DN)
```

Any info on the timeline for this to be pushed into the terraform official registry?

**Comment by lhercot** [comment]
Hi @edudppaz, Thanks for the analysis.

Most likely before Monday.

**Comment by edudppaz** [comment]
Hei. @lhercot  any info about when will this fix be released? Thanks.

**Comment by lhercot** [comment]
Sorry for the delay. I just released v0.10.0.
